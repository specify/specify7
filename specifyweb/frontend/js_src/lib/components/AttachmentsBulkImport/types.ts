import type React from 'react';
import type { LocalizedString } from 'typesafe-i18n';
import type { State } from 'typesafe-reducer';

import type { RA, RR } from '../../utils/types';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { Attachment, Tables } from '../DataModel/types';
import type { DatasetBase, DatasetBriefBase } from '../WbPlanView/Wrapped';
import type { PartialAttachmentUploadSpec } from './Import';
import type { staticAttachmentImportPaths } from './importPaths';
import type { keyLocalizationMapAttachment } from './utils';

export type UploadAttachmentSpec = {
  readonly token: string;
  readonly attachmentLocation: string;
};

type Matched = State<
  'matched',
  {
    readonly id: number;
  }
>;

type Skipped = State<
  'cancelled' | 'skipped',
  {
    readonly reason: keyof typeof keyLocalizationMapAttachment;
  }
>;
type Success = State<
  'success',
  {
    readonly successType: 'deleted' | 'uploaded';
  }
>;
export type AttachmentStatus = Matched | Skipped | Success;

export type PartialUploadableFileSpec = Partial<UploadableFileSpec> &
  Pick<UploadableFileSpec, 'uploadFile'>;

type UploadableFileSpec = {
  readonly uploadFile: UnBoundFile;
  readonly matchedId: RA<number>;
  readonly status: AttachmentStatus;
  readonly disambiguated: number;
  readonly attachmentId: number;
  readonly uploadTokenSpec: UploadAttachmentSpec;
  /*
   * This is added because if an attachment was uploaded correctly, but a network error
   * happened while saving the resource, and we retry, we will loose the previous attachment
   */
  readonly attachmentFromPreviousTry?: SpecifyResource<Attachment>;
};

/*
 * Forcing keys to be primitive because objects would be
 * ignored during syncing to backend.
 */
export type BoundFile = Pick<File, 'name' | 'size' | 'type'>;

export type UnBoundFile = {
  readonly file: BoundFile | File;
  readonly parsedName?: string;
};

export type AttachmentWorkProgress = {
  readonly total: number;
  readonly uploaded: number;
  readonly type: 'interrupted' | 'safe' | 'stopped' | 'stopping';
  readonly retryingIn: number;
  readonly stoppedByUser: boolean;
};

export type AttachmentWorkRef = {
  // eslint-disable-next-line functional/prefer-readonly-type
  mappedFiles: RA<PartialUploadableFileSpec>;
  // eslint-disable-next-line functional/prefer-readonly-type
  uploadPromise: Promise<number | undefined>;
  readonly retrySpec: Record<number, number>;
};

export type AttachmentWorkStateProps = {
  readonly onCompletedWork: (
    uploadables: RA<PartialUploadableFileSpec> | undefined
  ) => void;
  readonly workProgress: AttachmentWorkProgress;
  readonly workRef: React.MutableRefObject<AttachmentWorkRef>;
  readonly onStop: () => void;
  readonly triggerNow: () => void;
  readonly dialogText: RR<
    'onAction' | 'onCancelled' | 'onCancelledDescription',
    LocalizedString
  >;
};

export type AttachmentDatasetBrief = DatasetBriefBase & {
  readonly uploaderstatus:
    | 'deleting'
    | 'deletingInterrupted'
    | 'main'
    | 'uploading'
    | 'uploadInterrupted'
    | 'validating';
};

export type AttachmentDataSetPlan = AttachmentDatasetBrief & {
  readonly uploadplan: {
    readonly staticPathKey: string;
  };
};

export type AttachmentDataSet = AttachmentDatasetBrief &
  DatasetBase & {
    readonly rows: RA<PartialUploadableFileSpec>;
    readonly uploadplan: PartialAttachmentUploadSpec;
    readonly uploadresult?: {
      readonly timestamp: string;
    };
  };

export type FetchedDataSet =
  | AttachmentDataSet &
      (
        | { readonly uploaderstatus: 'main' }
        | ({
            readonly uploaderstatus: 'deleting' | 'uploading';
            readonly rows: RA<PartialUploadableFileSpec>;
          } & {
            readonly uploadplan: {
              readonly staticPathKey: keyof typeof staticAttachmentImportPaths;
            };
          })
      );

export type WrappedActionProps<KEY extends keyof Tables> = {
  readonly uploadableFile: PartialUploadableFileSpec;
  readonly baseTableName: KEY;
  readonly dryRun: boolean;
  readonly triggerRetry?: () => void;
};
