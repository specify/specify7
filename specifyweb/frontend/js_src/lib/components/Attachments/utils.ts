import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
