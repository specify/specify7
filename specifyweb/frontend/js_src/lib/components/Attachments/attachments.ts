import { commonText } from '../../localization/common';
import { ajax } from '../../utils/ajax';
import { Http } from '../../utils/ajax/definitions';
import { handleAjaxResponse } from '../../utils/ajax/response';
import type { IR, RA } from '../../utils/types';
import type { UploadAttachmentSpec } from '../AttachmentsBulkImport/types';
import { getField } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { tables } from '../DataModel/tables';
import type { Attachment } from '../DataModel/types';
import { load } from '../InitialContext';
import { getIcon, unknownIcon } from '../InitialContext/icons';
import { getPref } from '../InitialContext/remotePrefs';
import { formatUrl } from '../Router/queryString';

type AttachmentSettings = {
  readonly collection: string;
  readonly delete: string;
  readonly getmetadata: string;
  readonly read: string;
  readonly testkey: string;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  readonly token_required_for_get: boolean;
  readonly write: string;
};

let settings: AttachmentSettings | undefined;
export const attachmentSettingsPromise = load<AttachmentSettings | IR<never>>(
  '/context/attachment_settings.json',
  'application/json'
).then((data) => {
  if (Object.keys(data).length > 0) settings = data as AttachmentSettings;
  return attachmentsAvailable();
});

export const attachmentsAvailable = (): boolean => typeof settings === 'object';

const thumbnailable = new Set([
  'image/jpeg',
  'image/gif',
  'image/png',
  'image/tiff',
  'application/pdf',
]);

function iconForMimeType(mimeType: string): {
  readonly alt: string;
  readonly src: string;
} {
  if (mimeType === 'text/plain')
    return { alt: 'text', src: getIcon('text') ?? unknownIcon };
  if (mimeType === 'text/html')
    return { alt: 'html', src: getIcon('html') ?? unknownIcon };

  const parts = mimeType.split('/');
  const type = parts[0];
  const subtype = parts[1];

  if (['audio', 'video', 'image', 'text'].includes(type))
    return { alt: type, src: getIcon(type) ?? unknownIcon };

  if (type === 'application') {
    const iconName = {
      pdf: 'pdf',
      'vnd.ms-excel': 'MSExcel',
      'vnd.ms-word': 'MSWord',
      'vnd.ms-powerpoint': 'MSPowerPoint',
    }[subtype];

    if (typeof iconName === 'string')
      return { alt: iconName, src: getIcon(iconName) ?? unknownIcon };
  }

  return { alt: commonText.unknown(), src: getIcon('unknown') ?? unknownIcon };
}

export const fetchAssetToken = async (
  fileName: string,
  silent: boolean = false
) =>
  ajax(formatUrl('/attachment_gw/get_token/', { fileName }), {
    headers: { Accept: 'text/plain' },
    errorMode: silent ? 'silent' : 'dismissible',
    expectedErrors: silent ? Object.values(Http) : [Http.OK],
  }).then(({ data, status }) => (status === Http.OK ? data : undefined));

const fetchToken = async (fileName: string): Promise<string | undefined> =>
  settings?.token_required_for_get === true
    ? fetchAssetToken(fileName)
    : Promise.resolve(undefined);

export type AttachmentThumbnail = {
  readonly src: string;
  readonly alt: string | undefined;
  readonly width: number;
  readonly height: number;
};

export async function fetchThumbnail(
  attachment: SerializedResource<Attachment>,
  scale = getPref('attachment.preview_size')
): Promise<AttachmentThumbnail | undefined> {
  const mimeType = attachment.mimeType ?? undefined;
  const thumbnail =
    mimeType === undefined || thumbnailable.has(mimeType)
      ? undefined
      : iconForMimeType(mimeType);

  // Display an icon for resources that don't have a custom thumbnail
  if (typeof thumbnail === 'object' && thumbnail?.src !== unknownIcon)
    return {
      ...thumbnail,
      width: scale,
      height: scale,
    };

  // Fetch a preview for resources that support thumbnail
  if (attachment.attachmentLocation === null || settings === undefined)
    return undefined;
  const token = await fetchToken(attachment.attachmentLocation);
  return {
    src: formatUrl(settings.read, {
      coll: settings.collection,
      type: 'T',
      fileName: attachment.attachmentLocation,
      scale,
      token,
    }),
    alt: attachment.attachmentLocation ?? undefined,
    width: scale,
    height: scale,
  };
}

export const formatAttachmentUrl = (
  attachment: SerializedResource<Attachment>,
  token: string | undefined
): string | undefined =>
  typeof settings === 'object'
    ? formatUrl(settings.read, {
        coll: settings.collection,
        type: 'O',
        fileName: attachment.attachmentLocation,
        downloadName: attachment.origFilename?.replace(/^.*[/\\]/u, ''),
        token,
      })
    : undefined;

export const fetchOriginalUrl = async (
  attachment: SerializedResource<Attachment>
): Promise<string | undefined> =>
  typeof attachment.attachmentLocation === 'string'
    ? fetchToken(attachment.attachmentLocation).then((token) =>
        formatAttachmentUrl(attachment, token)
      )
    : Promise.resolve(undefined);

export async function uploadFile(
  file: File,
  handleProgress: (percentage: number | true) => void,
  uploadAttachmentSpec?: UploadAttachmentSpec,
  strict = true
): Promise<SpecifyResource<Attachment> | undefined> {
  if (settings === undefined) return undefined;

  const data =
    typeof uploadAttachmentSpec === 'object'
      ? uploadAttachmentSpec
      : await ajax<RA<Partial<UploadAttachmentSpec>>>(
          '/attachment_gw/get_upload_params/',
          {
            method: 'POST',
            headers: { Accept: 'application/json' },
            body: {
              filenames: [file.name],
            },
          }
        ).then(({ data }) => data[0]);

  if (data.attachmentLocation === undefined || data.token === undefined)
    return undefined;

  const formData = new FormData();

  formData.append('file', file);
  formData.append('token', data.token);
  formData.append('store', data.attachmentLocation);
  formData.append('type', 'O');
  formData.append('coll', settings.collection);

  /*
   * Using XMLHttpRequest rather than fetch() because need upload
   * progress reporting, which is not yet supported by fetch API
   */

  const xhr = new XMLHttpRequest();
  xhr.upload?.addEventListener('progress', (event) =>
    handleProgress(event.lengthComputable ? event.loaded / event.total : true)
  );
  xhr.open('POST', settings.write);
  xhr.send(formData);
  const DONE = 4;
  await new Promise((resolve, reject) =>
    xhr.addEventListener('readystatechange', () => {
      if (xhr.readyState === DONE)
        try {
          resolve(
            handleAjaxResponse({
              expectedErrors: [],
              accept: undefined,
              // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
              response: {
                ok: xhr.status === Http.OK,
                status: xhr.status,
                url: settings!.write,
              } as Response,

              errorMode: strict ? 'visible' : 'silent',
              text: xhr.responseText,
            })
          );
        } catch (error) {
          reject(error);
        }
    })
  );
  return new tables.Attachment.Resource({
    attachmentlocation: data.attachmentLocation,
    mimetype: fixMimeType(file.type),
    origfilename: file.name,
    title: file.name,
    isPublic: getPref('attachment.is_public_default'),
  });
}

/**
 * A temporary workaround for mimeTypes for `.docx` and `.xlsx` files being
 * longer than the length limit on the `Attachment.mimeType` field.
 * See: https://github.com/specify/specify7/issues/1141
 * REFACTOR: remove this once that issue is fixed
 */
function fixMimeType(originalMimeType: string): string {
  const maxLength = getField(tables.Attachment, 'mimeType').length;
  if (maxLength === undefined || originalMimeType.length < maxLength)
    return originalMimeType;
  else {
    const mimeType = 'application/octet-stream';
    console.warn(
      `Shortened the Attachment mimeType from "${originalMimeType}"` +
        ` to "${mimeType}" due to length limit`
    );
    return mimeType;
  }
}
