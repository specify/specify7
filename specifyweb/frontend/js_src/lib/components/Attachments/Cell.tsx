import React from 'react';

import { useBooleanState } from '../../hooks/useBooleanState';
import { attachmentsText } from '../../localization/attachments';
import { commonText } from '../../localization/common';
import { f } from '../../utils/functools';
import type { GetSet } from '../../utils/types';
import { Button } from '../Atoms/Button';
import { LoadingContext } from '../Core/Contexts';
import { fetchRelated } from '../DataModel/collection';
import type { AnySchema, SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { idFromUrl } from '../DataModel/resource';
import { deserializeResource } from '../DataModel/serializers';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { getTableById } from '../DataModel/tables';
import type { Attachment } from '../DataModel/types';
import { softFail } from '../Errors/Crash';
import { Dialog } from '../Molecules/Dialog';
import { TableIcon } from '../Molecules/TableIcon';
import { hasTablePermission } from '../Permissions/helpers';
import { AttachmentPreview } from './Preview';
import { getAttachmentRelationship, tablesWithAttachments } from './utils';

export function AttachmentCell({
  attachment,
  onOpen: handleOpen,
  related: [related, setRelated],
  onViewRecord: handleViewRecord,
}: {
  readonly attachment: SerializedResource<Attachment>;
  readonly onOpen: () => void;
  readonly related: GetSet<SpecifyResource<AnySchema> | undefined>;
  readonly onViewRecord:
    | ((table: SpecifyTable, recordId: number) => void)
    | undefined;
}): JSX.Element {
  const table = f.maybe(attachment.tableID ?? undefined, getAttachmentTable);

  return (
    <div className="relative">
      {typeof handleViewRecord === 'function' &&
      table !== undefined &&
      hasTablePermission(table.name, 'read') ? (
        <AttachmentRecordLink
          attachment={attachment}
          className="absolute left-0 top-0"
          related={[related, setRelated]}
          table={table}
          variant="icon"
          onViewRecord={handleViewRecord}
        />
      ) : undefined}
      <AttachmentPreview
        attachment={attachment}
        onOpen={(): void => {
          if (related === undefined && typeof table === 'object')
            fetchAttachmentParent(table, attachment)
              .then(setRelated)
              .catch(softFail);
          handleOpen();
        }}
      />
    </div>
  );
}

export function getAttachmentTable(tableId: number): SpecifyTable | undefined {
  const table = getTableById(tableId);
  return tablesWithAttachments().includes(table) ? table : undefined;
}

/**
 * A button to open a record associated with the attachment
 */
export function AttachmentRecordLink({
  variant,
  className,
  table,
  attachment,
  onViewRecord: handleViewRecord,
  related: [related, setRelated],
}: {
  readonly variant: 'button' | 'icon';
  readonly className: string;
  readonly table: SpecifyTable;
  readonly attachment: SerializedResource<Attachment>;
  readonly onViewRecord: (table: SpecifyTable, recordId: number) => void;
  readonly related: GetSet<SpecifyResource<AnySchema> | undefined>;
}): JSX.Element {
  const loading = React.useContext(LoadingContext);
  const [isFailed, handleFailed, handleNotFailed] = useBooleanState();
  const Component = variant === 'icon' ? Button.LikeLink : Button.Info;
  return (
    <>
      <Component
        className={className}
        title={table?.label}
        onClick={(): void =>
          loading(
            (typeof related === 'object'
              ? Promise.resolve(related)
              : fetchAttachmentParent(table, attachment).then((related) => {
                  setRelated(related);
                  return related;
                })
            )
              .then((related) =>
                typeof related === 'object'
                  ? getBaseResourceId(table, related)
                  : undefined
              )
              .then((id) =>
                typeof id === 'number'
                  ? handleViewRecord(table, id)
                  : handleFailed()
              )
          )
        }
      >
        <TableIcon label name={table?.name ?? 'Attachment'} />
        {variant === 'button' && table?.label}
      </Component>
      {isFailed ? (
        <Dialog
          buttons={commonText.close()}
          header={attachmentsText.unableToFindRelatedRecord()}
          onClose={handleNotFailed}
        >
          {attachmentsText.unableToFindRelatedRecordDescription()}
        </Dialog>
      ) : undefined}
    </>
  );
}

/** Fetch CollectionObjectAttachment for a given Attachment */
async function fetchAttachmentParent(
  table: SpecifyTable,
  attachment: SerializedResource<Attachment>
): Promise<SpecifyResource<AnySchema> | undefined> {
  const { records } = await fetchRelated(
    attachment,
    getAttachmentRelationship(table)!.name as 'collectionObjectAttachments'
  );
  return deserializeResource(records[0]);
}

/**
 * Get CollectionObject id from CollectionObjectAttachment
 */
function getBaseResourceId(
  table: SpecifyTable,
  related: SpecifyResource<AnySchema>
): number | undefined {
  // This would be a URL to CollectionObject
  const resourceUrl = related.get(table.name as 'CollectionObject');
  return idFromUrl(resourceUrl ?? '');
}
