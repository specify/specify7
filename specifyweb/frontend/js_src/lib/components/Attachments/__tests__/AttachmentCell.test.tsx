import { waitFor } from "@testing-library/react";
import React from "react";

import * as AsyncState from "../../../hooks/useAsyncState";
import { clearIdStore } from "../../../hooks/useId";
import { overrideAjax } from "../../../tests/ajax";
import attachmentSettings from '../../../tests/ajax/static/context/attachment_settings.json';
import { requireContext } from "../../../tests/helpers";
import { mount } from "../../../tests/reactUtils";
import { f } from "../../../utils/functools";
import { type GetOrSet, type GetSet, overwriteReadOnly } from "../../../utils/types";
import { LoadingContext } from "../../Core/Contexts";
import type { AnySchema, SerializedResource } from "../../DataModel/helperTypes";
import type { SpecifyResource } from "../../DataModel/legacyTypes";
import type { SpecifyTable } from "../../DataModel/specifyTable";
import { tables } from "../../DataModel/tables";
import type { Attachment } from "../../DataModel/types";
import { overrideAttachmentSettings } from "../attachments";
import { AttachmentCell } from "../Cell";
import { testAttachment, testCollectionObjectAttachment } from "./utils";

requireContext();

const setStateMock = jest.fn();

beforeEach(() => {
    setStateMock.mockClear();
    clearIdStore();
});

jest.spyOn(AsyncState, 'useAsyncState').mockImplementation(useAsyncRef);

function useAsyncRef<T>(
    callback: () => Promise<T | undefined> | T | undefined,
    _: boolean
): GetOrSet<T | undefined> {
    const ref = React.useRef<T | undefined>(undefined);

    React.useLayoutEffect(() => {
        ref.current = undefined;

        Promise.resolve(callback()).then((data) => {
            if (destructorCalled) return;
            ref.current = data;
            setStateMock(data);
        });

        let destructorCalled = false;

        return () => {
            destructorCalled = true;
        }
    }, [callback]);

    const setCallback = (...value: Parameters<GetOrSet<T | undefined>[1]>) => {
        const setter = value[0];
        ref.current = typeof setter === 'function' ? (setter as Function)(ref.current) : setter;
    }

    return [ref.current, setCallback];
}


function AttachmentCellMock({ options, ...rest }: {
    readonly options: {
        readonly trigger: () => void,
    }
    readonly attachment: SerializedResource<Attachment>;
    readonly onOpen: () => void;
    readonly related: GetSet<SpecifyResource<AnySchema> | undefined>;
    readonly onViewRecord:
    | ((table: SpecifyTable, recordId: number) => void)
    | undefined;
}) {
    const [_, setState] = React.useState<boolean>(false);
    const triggerChange = React.useCallback(() => {
        setState((old) => !old);
    }, [setState]);
    overwriteReadOnly(options, 'trigger', triggerChange);
    return (<AttachmentCell {...rest} />)

}
describe("AttachmentCell", () => {

    overrideAjax("/api/specify/collectionobjectattachment/?limit=20&attachment=5490",
        {
            "objects": [
                testCollectionObjectAttachment
            ],
            "meta": {
                "limit": 20,
                "offset": 0,
                "total_count": 1
            }
        })


    test("simple render", async () => {

        jest.spyOn(console, 'warn').mockImplementation();

        const handleOpen = jest.fn();
        const setRelated = jest.fn();
        const handleViewRecord = jest.fn();
        const options = { trigger: f.never }
        overrideAttachmentSettings(attachmentSettings);

        const { asFragment, rerender, getAllByRole, user } = mount(
            <LoadingContext.Provider value={f.void} >
                <AttachmentCellMock
                    attachment={testAttachment}
                    options={options}
                    related={[undefined, setRelated]}
                    onOpen={handleOpen}
                    onViewRecord={handleViewRecord}
                />
            </LoadingContext.Provider>
        );

        await waitFor(() => { expect(setStateMock).toHaveBeenCalled() });

        rerender(
            <LoadingContext.Provider value={f.void} >
                <AttachmentCellMock
                    attachment={testAttachment}
                    options={options}
                    related={[undefined, setRelated]}
                    onOpen={handleOpen}
                    onViewRecord={handleViewRecord}
                />
            </LoadingContext.Provider>
        );

        expect(asFragment()).toMatchSnapshot();

        const button = getAllByRole('button')[1];

        expect(handleOpen).not.toHaveBeenCalled();
        await user.click(button)

        expect(handleOpen).toHaveBeenCalled();

        const relatedButton = getAllByRole('button')[0];
        await user.click(relatedButton);
        expect(setRelated.mock.lastCall).toMatchSnapshot();

        expect(handleViewRecord).toHaveBeenCalled();
        expect(handleViewRecord).toHaveBeenCalledWith(tables.CollectionObject, 51_731);

    });
})