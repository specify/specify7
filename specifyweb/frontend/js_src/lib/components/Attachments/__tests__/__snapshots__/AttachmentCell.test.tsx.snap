// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentCell simple render 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <button
      class="link absolute left-0 top-0"
      title="Collection Object"
      type="button"
    >
      <svg
        aria-hidden="false"
        class="w-table-icon h-table-icon flex-shrink-0 font-normal not-italic"
        role="img"
        viewBox="0 0 1000 1000"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Collection Object
        </title>
        <g>
          <radialgradient
            cx="512.8199"
            cy="2.0972"
            gradientUnits="userSpaceOnUse"
            id="icon-0-gradient"
            r="1190.8768"
          >
            <stop
              offset="0"
              style="stop-color: rgb(166, 124, 82);"
            />
            <stop
              offset="1"
              style="stop-color: rgb(117, 76, 36);"
            />
          </radialgradient>
          <path
            d="M500,0c53.8,0.7,107.6,0.9,161.4,2.4c51.6,1.4,103,5.1,153.2,18.4c57.7,15.3,104.7,45.7,136.6,97.1c19.9,32.1,30.1,67.7,36.5,104.5c9.3,53.9,11.4,108.3,11.6,162.8c0.3,92.3,0,184.7-1.3,277c-0.7,49.8-4.8,99.5-17,148.1c-15.2,60.4-46.1,109.8-100.4,142.6c-31.5,19.1-66.2,29.1-102.2,35.2c-60.2,10.3-121.1,11.8-182,11.9c-85.9,0.1-171.8-0.2-257.7-1.5c-49.1-0.7-98.1-4.8-146.1-16.6C97.9,958.9,40.1,899.6,17.8,805.2C9,768,4.8,730.1,3.2,692.1c-5-119.8-4.1-239.6-1.1-359.3c1.3-50.9,5.5-101.7,19.4-151.1c24.6-87.2,81.8-140.4,169-162.3C239.1,7.2,288.8,3.6,338.6,2.3C392.4,0.9,446.2,0.7,500,0z"
            fill="url(#icon-0-gradient)"
          />
        </g>
        <g>
          <text
            alignment-baseline="middle"
            baseline-shift="-10%"
            dominant-baseline="central"
            fill="#FFFFFF"
            font-family="Francois One,sans-serif"
            font-size="660px"
            text-anchor="middle"
            x="50%"
            y="50%"
          >
            CO
          </text>
        </g>
      </svg>
    </button>
    <div
      class="
        flex items-center justify-center rounded bg-white shadow-md
        shadow-gray-300 transition hover:shadow-md hover:shadow-gray-400
        dark:bg-black
      "
    >
      <button
        type="button"
      >
        <img
          alt="kubiodiversitylogo.jpg"
          class="
        max-h-full max-w-full border-2 border-white object-contain
        dark:border-black
      "
          src="http://host.docker.internal/fileget?coll=KU+Fish+Voucher+Collection&type=T&filename=a6daad6c-5091-4461-86ee-eeba9434ef2b.jpg&scale=123"
          style="width: 123px; height: 123px;"
        />
      </button>
    </div>
    <a
      aria-label="Download"
      class="icon link rounded absolute right-0 top-0"
      download="topbrc4.jpg"
      href="/attachment_gw/proxy/?coll=KU+Fish+Voucher+Collection&type=O&filename=a6daad6c-5091-4461-86ee-eeba9434ef2b.jpg&downloadname=topbrc4.jpg"
      target="_blank"
      title="Download"
    >
      <svg
        aria-hidden="true"
        class="w-6 h-6 flex-shrink-0"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 17a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm3.293-7.707a1 1 0 011.414 0L9 10.586V3a1 1 0 112 0v7.586l1.293-1.293a1 1 0 111.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </a>
  </div>
</DocumentFragment>
`;

exports[`AttachmentCell simple render 2`] = `
[
  {
    "attachment": {
      "accessionattachments": "/api/specify/accessionattachment/?attachment=5490",
      "agentattachments": "/api/specify/agentattachment/?attachment=5490",
      "attachmentimageattribute": null,
      "attachmentlocation": "a6daad6c-5091-4461-86ee-eeba9434ef2b.jpg",
      "attachmentstorageconfig": null,
      "borrowattachments": "/api/specify/borrowattachment/?attachment=5490",
      "capturedevice": null,
      "collectingeventattachments": "/api/specify/collectingeventattachment/?attachment=5490",
      "collectingtripattachments": "/api/specify/collectingtripattachment/?attachment=5490",
      "collectionobjectattachments": "/api/specify/collectionobjectattachment/?attachment=5490",
      "conservdescriptionattachments": "/api/specify/conservdescriptionattachment/?attachment=5490",
      "conserveventattachments": "/api/specify/conserveventattachment/?attachment=5490",
      "copyrightdate": null,
      "copyrightholder": null,
      "createdbyagent": "/api/specify/agent/3/",
      "creator": null,
      "credit": null,
      "dateimaged": null,
      "deaccessionattachments": "/api/specify/deaccessionattachment/?attachment=5490",
      "disposalattachments": "/api/specify/disposalattachment/?attachment=5490",
      "dnasequenceattachments": "/api/specify/dnasequenceattachment/?attachment=5490",
      "dnasequencingrunattachments": "/api/specify/dnasequencingrunattachment/?attachment=5490",
      "exchangeinattachments": "/api/specify/exchangeinattachment/?attachment=5490",
      "exchangeoutattachments": "/api/specify/exchangeoutattachment/?attachment=5490",
      "fieldnotebookattachments": "/api/specify/fieldnotebookattachment/?attachment=5490",
      "fieldnotebookpageattachments": "/api/specify/fieldnotebookpageattachment/?attachment=5490",
      "fieldnotebookpagesetattachments": "/api/specify/fieldnotebookpagesetattachment/?attachment=5490",
      "filecreateddate": null,
      "giftattachments": "/api/specify/giftattachment/?attachment=5490",
      "guid": "7664fc1b-cc39-4f7d-b81c-645fc11e8e1d",
      "id": 5490,
      "ispublic": true,
      "license": null,
      "licenselogourl": null,
      "loanattachments": "/api/specify/loanattachment/?attachment=5490",
      "localityattachments": "/api/specify/localityattachment/?attachment=5490",
      "metadata": "/api/specify/attachmentmetadata/?attachment=5490",
      "metadatatext": null,
      "mimetype": "image/jpeg",
      "modifiedbyagent": null,
      "origfilename": "topbrc4.jpg",
      "permitattachments": "/api/specify/permitattachment/?attachment=5490",
      "preparationattachments": "/api/specify/preparationattachment/?attachment=5490",
      "referenceworkattachments": "/api/specify/referenceworkattachment/?attachment=5490",
      "remarks": null,
      "repositoryagreementattachments": "/api/specify/repositoryagreementattachment/?attachment=5490",
      "resource_uri": "/api/specify/attachment/5490/",
      "scopeid": null,
      "scopetype": null,
      "spdatasetattachments": "/api/specify/spdatasetattachment/?attachment=5490",
      "storageattachments": "/api/specify/storageattachment/?attachment=5490",
      "subjectorientation": null,
      "subtype": null,
      "tableid": 1,
      "tags": "/api/specify/attachmenttag/?attachment=5490",
      "taxonattachments": "/api/specify/taxonattachment/?attachment=5490",
      "timestampcreated": "2015-05-20T12:16:38",
      "timestampmodified": "2015-05-20T12:16:38",
      "title": "kubiodiversitylogo.jpg",
      "treatmenteventattachments": "/api/specify/treatmenteventattachment/?attachment=5490",
      "type": null,
      "version": 0,
      "visibility": null,
      "visibilitysetby": null,
    },
    "collectionmemberid": 4,
    "collectionobject": "/api/specify/collectionobject/51731/",
    "createdbyagent": "/api/specify/agent/3/",
    "id": 4732,
    "modifiedbyagent": null,
    "ordinal": 0,
    "remarks": null,
    "resource_uri": "/api/specify/collectionobjectattachment/4732/",
    "timestampcreated": "2025-07-31T17:00:40",
    "timestampmodified": "2025-07-31T17:00:40",
    "version": 0,
  },
]
`;
